File: src/java/3_StructuralPatterns/5_Flyweight/FlyweightExercise.java
package com.activemesa.structural.flyweight.exercise;

import java.util.*;

class Sentence
{
  private String [] words;
  private Map<Integer, WordToken> tokens = new HashMap<>();

  public Sentence(String plainText)
  {
    words = plainText.split(" ");
  }

  public WordToken getWord(int index)
  {
    WordToken wt = new WordToken();
    tokens.put(index, wt);
    return tokens.get(index);
  }

  @Override
  public String toString()
  {
    ArrayList<String> ws = new ArrayList<>();
    for (int i = 0; i < words.length; ++i)
    {
      String w = words[i];
      if (tokens.containsKey(i) && tokens.get(i).capitalize)
        w = w.toUpperCase();
      ws.add(w);
    }
    return String.join(" ", ws);
  }

  class WordToken
  {
    public boolean capitalize;
  }
}

File: src/js/3_StructuralPatterns/6_Flyweight/flyweight_exercise.js
class WordToken
{
  constructor(capitalize=false)
  {
    this.capitalize = capitalize;
  }
}

class Sentence
{
  constructor(plainText)
  {
    this.words = plainText.split(' ');
    this.tokens = {};
  }

  at(index)
  {
    let wt = new WordToken();
    this.tokens[index] = wt;
    return this.tokens[index];
  }

  toString()
  {
    let buffer = [];
    for (let i = 0; i < this.words.length; ++i)
    {
      let w = this.words[i];
      if (this.tokens[i] && this.tokens[i].capitalize)
        w = w.toUpperCase();
      buffer.push(w);
    }
    return buffer.join(' ');
  }
}

describe('flyweight', function()
{
  it('should let us capitalize words', function()
  {
    let s = new Sentence('alpha beta gamma');
    s.at(1).capitalize = true;
    expect(s.toString()).toEqual('alpha BETA gamma');
  });
});